Disable Copy as Linked Application when appropriate

gherceg opened the pull request:
## Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SS-173

This is a fun one. I'm currently working on a solution to link apps directly from the linked domains UI, but in the meantime I need to update the app manager linking apps UI to support the `RELEASE_MANAGEMENT` privilege version. I've assigned a medium risk here because this has the potential to impact copying apps which is a more generally used feature than linked domains. The goal here is the following:

- prevent users from being able to link _domains_ via this interface (don't want them to pick up "bad" habits)
- only allow admins the ability to link apps from this interface
- do not disrupt regular copying of apps

This can be achieved by controlling when the `Copy as Linked Application` checkbox is enabled. In the `RELEASE_MANAGEMENT` version, a user should only be able to link an app in an **existing** domain link. So by supplying a list of active downstream domains that the current user is an admin in, we can check each time a domain is selected from the dropdown if the selected domain is an active downstream domain. If it is, we can enable the linked app checkbox. If it isn't, we disable and **make sure it is unchecked**. 

Here are some screenshots of this in action. 

<img width="759" alt="Screen Shot 2021-07-23 at 5 27 42 PM" src="https://user-images.githubusercontent.com/15785053/126853742-5df08e3d-1fa4-472c-a8ea-5edc1dc43a19.png">

<img width="804" alt="Screen Shot 2021-07-22 at 4 20 17 PM" src="https://user-images.githubusercontent.com/15785053/126740210-218ccefc-6deb-4940-8fd2-05b49323c3cf.png">

P.S. - sorry for refactoring a method name that is used in a decent number of files.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
For domains with the `RELEASE_MANAGEMENT` privilege, they will be limited on which domains they can copy an app as a linked application to.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
No automated tests were added. This was primarily a UI change.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
Definitely will run through QA on the parent branch.
### Safety story
<!--
Describe any other pieces to the safety story including
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
Locally tested.
### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

gherceg committed:
disable checkbox if necessary

- if RELEASE_MANAGEMENT privilege enabled, only able to link apps with active
downstream domains
- behavior should remain the same for LINKED_DOMAINS toggle users

gherceg committed:
keep same behavior as before for linked_domains users

gherceg committed:
enforce admin for RELEASE_MANAGEMENT version

gherceg committed:
add user :facepalm:

gherceg committed:
should have used request.couch_user

orangejenny commented on corehq/apps/app_manager/forms.py:
```diff
@@ -18,20 +19,31 @@
+        widget=BootstrapCheckboxInput(
+            inline_label=mark_safe(_(
+                "<!-- ko ifnot: shouldEnableLinkedAppOption -->"
+                "Cannot link this application with the selected project space."
```
Could this say **why** app can't be linked?

gherceg replied:
Yeah I thought about this. I kept it vague because I thought about it as 2 separate reasons. One being that the domain isn't part of the enterprise account, and the other being it is part of the enterprise account but isn't link. In hindsight, saying the domain isn't linked is sufficient, and they can figure out via the domain linking interface/documentation that they can only link to domains within the enterprise account. 

I'll change it to "The selected project space must first be linked to the current project space." for now, and bring this up to product.

orangejenny commented on corehq/apps/linked_domain/dbaccessors.py:
```diff
@@ -98,3 +98,19 @@ def _is_available_upstream_domain(candidate_name):
+def get_domains_eligible_for_linked_apps(upstream_domain_name, user):
+    if domain_has_privilege(upstream_domain_name, RELEASE_MANAGEMENT):
+        upstream_membership = user.get_domain_membership(upstream_domain_name)
+        is_upstream_admin = upstream_membership.is_admin if upstream_membership else False
```
Should this also check for superusers, who won't have a membership?

gherceg replied:
Ah I didn't think about that. That does sound reasonable, and I should also make that change to the method that retrieves domains that are available to link, `get_available_domains_to_link`.

orangejenny commented:
> sorry for refactoring a method name that is used in a decent number of files

it's cool, the new name is better

gherceg committed:
check for superuser access

gherceg committed:
break up method to get available domains to link

gherceg commented:
> Should this also check for superusers, who won't have a membership?

So as I tried to figure out the best way to test superuser access, I realized the current implementation of `get_available_domains_to_link` was turning into a scrambled mess and I needed to break it up. The [latest commit](https://github.com/dimagi/commcare-hq/pull/30116/commits/f7cb58cb58ae882a901844f1b662e70a73683d6c) is essentially a giant refactor of that method to make it more readable and testable. I replaced the current test case for that method with a simple test case that relied on mocks, and relied more on testing the underlying units of that method. I would recommend looking at that last commit on it's own.

gherceg committed:
update message text

gherceg committed:
Merge branch 'gh/base-linked-project-access' into gh/limit-domains-in-app-manager

gherceg committed:
error in tests

orangejenny commented on corehq/apps/app_manager/forms.py:
```diff
@@ -37,7 +37,7 @@ class CopyApplicationForm(forms.Form):
             inline_label=mark_safe(_(
                 "<!-- ko ifnot: shouldEnableLinkedAppOption -->"
-                "Cannot link this application with the selected project space."
+                "The selected project space must first be linked to the current project space."
```
👍

orangejenny commented on corehq/apps/linked_domain/util.py:
```diff
@@ -96,3 +96,34 @@ def _add_domain_access(domain, media):
+
+
+def user_has_admin_access_in_all_domains(user, domains):
+    if user.is_superuser:
```
[is_domain_admin](https://github.com/dimagi/commcare-hq/blob/18cacfb4e6dfd7dc77d3ba3c281f90f39606c33b/corehq/apps/users/models.py#L658) is sufficient here - it already checks is_superuser, and it also respects the `domain_restricts_superusers` options.

gherceg replied:
🤦 yeah that should be sufficient. As I was writing this, I felt like there must already be something that exists like this. Thanks for the tip.

gherceg committed:
is_domain_admin already checks for superuser; remove tests

orangejenny commented on corehq/apps/linked_domain/dbaccessors.py:
```diff
@@ -50,43 +53,42 @@ def get_actions_in_domain_link_history(link):
     version is available to anyone who can obtain access (the wild west)
     :param upstream_domain_name: potential upstream domain candidate
     :param user: user object
+    :param billing_account: optional parameter used
```
```suggestion
    :param billing_account: optional parameter used to limit available domains to a specific enterprise account
```

orangejenny commented on corehq/apps/linked_domain/views.py:
```diff
@@ -278,7 +278,10 @@ def page_context(self):
         )
 
-        available_domains_to_link = get_available_domains_to_link(self.request.domain, self.request.couch_user)
+        current_subscription = Subscription.get_active_subscription_by_domain(self.request.domain)
```
FYI there's a `BillingAccount.get_account_by_domain` does this same logic.

gherceg replied:
good point

gherceg committed:
address Jenny's feedback

- update documentation for billing_account
- use BillingAccount.get_account_by_domain instead
